File: packages/core/src/marginalia_core/domain/note.py
"""Voice note domain models."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from pathlib import Path

from marginalia_core.domain.reading_session import ReadingPosition


def _utc_now() -> datetime:
    return datetime.now(UTC)


@dataclass(frozen=True, slots=True)
class VoiceNote:
    """A note anchored to a specific reading position."""

    note_id: str
    session_id: str
    document_id: str
    position: ReadingPosition
    transcript: str
    raw_audio_path: Path | None = None
    created_at: datetime = field(default_factory=_utc_now)

    @property
    def anchor(self) -> str:
        return self.position.anchor

File: crates/marginalia-core/src/domain/note.rs
use crate::domain::reading_session::ReadingPosition;
use chrono::{DateTime, Utc};
use std::path::PathBuf;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VoiceNote {
    pub note_id: String,
    pub session_id: String,
    pub document_id: String,
    pub position: ReadingPosition,
    pub transcript: String,
    pub transcription_provider: String,
    pub language: String,
    pub raw_audio_path: Option<PathBuf>,
    pub created_at: DateTime<Utc>,
}

impl VoiceNote {
    pub fn anchor(&self) -> String {
        self.position.anchor()
    }
}

#[cfg(test)]
mod tests {
    use super::VoiceNote;
    use crate::domain::reading_session::ReadingPosition;
    use chrono::Utc;

    #[test]
    fn voice_note_anchor_comes_from_position() {
        let note = VoiceNote {
            note_id: "note-1".to_string(),
            session_id: "session-1".to_string(),
            document_id: "doc-1".to_string(),
            position: ReadingPosition {
                section_index: 1,
                chunk_index: 4,
                char_offset: 0,
            },
            transcript: "Important thought".to_string(),
            transcription_provider: "fake".to_string(),
            language: "it".to_string(),
            raw_audio_path: None,
            created_at: Utc::now(),
        };

        assert_eq!(note.anchor(), "section:1/chunk:4");
    }
}
